tests/js/client/shell/api/cursor.js

```diff
@@ -1212,9 +1212,268 @@ function dealing_with_cursorsSuite_fetching_floating_point_valuesSuite () {
+      assertTrue(doc.parsedBody['hasMore']);
+      assertEqual(doc.parsedBody['result'].length, 1000);
+      assertFalse(doc.parsedBody['cached']);
+
+      let cursorId = doc.parsedBody['id'];
+      assertEqual(doc.parsedBody['nextBatchId'], undefined);
+
+      internal.debugSetFailAt("MakeConnectionErrorForRetry");
+
+
+      cmd = api + `/${cursorId}`;
+      doc = arango.POST_RAW(cmd, "");
+
+      assertEqual(doc.code, 500);
+      assertTrue(doc.error);
+
+      internal.debugClearFailAt();
+      // as the `batchId` field starts with 1 for the first batch, we assume the next batch id would be 2 for
+      // the failing request to send another request to retrieve its result, knowing it would fail because
+      // the flag `retriable`is false
+      cmd = api + `/${cursorId}/2`;
+      doc = arango.POST_RAW(cmd, "");
+      assertEqual(doc.code, 400);
+      assertEqual(doc.headers['content-type'], contentType);
+      assertTrue(doc.parsedBody['error']);
+      assertEqual(doc.parsedBody['code'], 400);
+      assertEqual(doc.parsedBody['errorMessage'], 'expecting retriable option to be true');
+    },
+
+    test_cursor_non_stream_retriable: function () {
+      const stmt = db._createStatement({
+        query: `FOR u IN ${cn} RETURN u`,
+        options: {stream: false, retriable: true},
+        batchSize: 100
+      });
+      let cursor = stmt.execute();
+      internal.debugSetFailAt("MakeConnectionErrorForRetry");
+
+      // the batches in between will also be retrieved with `/_api/cursor/<cursorId>/<latestBatchId>` because
+      // the failure point is located in a place in which the server would return an error, hence not returning
+      // the batch response object to the user, but it would have been constructed, so the latest batch would be cached
+      for (let i = 0; i < 2001; ++i) {
+        const nextValue = cursor.next();
+        // last batch not returning will close the cursor, won't be able to fetch the latest batch
+        if (i === 1900) {
+          internal.debugClearFailAt();
+        }
+        assertEqual("test" + i, nextValue._key);
+        assertEqual(i !== 2000, cursor.hasNext());
+      }
+      assertFalse(cursor.hasNext());
+    },
+
+    test_cursor_stream_retriable: function () {
+      const stmt = db._createStatement({
+        query: `FOR u IN ${cn} RETURN u`,
+        options: {stream: true, retriable: true},
+        batchSize: 100
+      });
+      let cursor = stmt.execute();
+      internal.debugSetFailAt("MakeConnectionErrorForRetry");
+
+      for (let i = 0; i < 2001; ++i) {
+        const nextValue = cursor.next();
+        // last batch not returning will close the cursor, won't be able to fetch the latest batch
+        if (i === 1900) {
+          internal.debugClearFailAt();
+        }
+        assertEqual("test" + i, nextValue._key);
+        assertEqual(i !== 2000, cursor.hasNext());
+      }
+      assertFalse(cursor.hasNext());
+    },
+  };
+}
```

Comment:
There should also be a test that tries to receive an invalid batch ids, e.g. when the current batch is is 1, it should try to retrieve batch id 2, or when the current id is 2, it should try to retrieve 0. 
Or try to receive batch id `"piff"` or such.

Reply:
Added within the same test functions, but expecting a 404 for getting the `nextBatchId` when it's one batch ahead (not yet fetched) and for batch id 0